_posts/2020-02-26-demo-fraud-detection-2.md

```diff
@@ -0,0 +1,210 @@
+
+```java
+public static final MapStateDescriptor<Integer, Rule> RULES_STATE_DESCRIPTOR =
+        new MapStateDescriptor<>("rules", Integer.class, Rule.class);
+```
+
+Connecting to `rulesStream` causes some changes in the signature of the processing functions. The previous article presented it in a slightly simplified way as a `ProcessFunction`. However, `DynamicKeyFunction` is actually a `BroadcastProcessFunction`.
+
+```java
+public abstract class BroadcastProcessFunction<IN1, IN2, OUT> {
+
+    public abstract void processElement(IN1 value,
+                                        ReadOnlyContext ctx,
+                                        Collector<OUT> out) throws Exception;
+
+    public abstract void processBroadcastElement(IN2 value,
+                                                 Context ctx,
+                                                 Collector<OUT> out) throws Exception;
+
+}
+```
+
+The difference is the addition of the `processBroadcastElement` method through which messages of the rules stream will arrive. The following new version of `DynamicKeyFunction` allows modifying the list of data-distribution keys at runtime through this stream:
+
+```java
+public class DynamicKeyFunction
+    extends BroadcastProcessFunction<Transaction, Rule, Keyed<Transaction, String, Integer>> {
+
+
+  @Override
+  public void processBroadcastElement(Rule rule,
+                                     Context ctx,
+                                     Collector<Keyed<Transaction, String, Integer>> out) {
+    BroadcastState<Integer, Rule> broadcastState = ctx.getBroadcastState(RULES_STATE_DESCRIPTOR);
+    broadcastState.put(rule.getRuleId(), rule);
+  }
+
+  @Override
+  public void processElement(Transaction event,
+                           ReadOnlyContext ctx,
+                           Collector<Keyed<Transaction, String, Integer>> out){
+    ReadOnlyBroadcastState<Integer, Rule> rulesState =
+                                  ctx.getBroadcastState(RULES_STATE_DESCRIPTOR);
+    for (Map.Entry<Integer, Rule> entry : rulesState.immutableEntries()) {
+        final Rule rule = entry.getValue();
+        out.collect(
+          new Keyed<>(
+            event, KeysExtractor.getKey(rule.getGroupingKeyNames(), event), rule.getRuleId()));
+    }
+  }
+}
+```
+
+In the above code, `processElement()` receives Transactions, and `processBroadcastElement()` receives Rules updates. When a new rule is created, it is distributed as depicted in Figure 6 and saved in all parallel instances of the operator using `processBroadcastState`. We use a Rule's ID as the key to store and reference individual rules. Instead of iterating over a hardcoded `List<Rules>`, we iterate over entries in the dynamically-updated broadcast state.
```

Comment:
```suggestion
In the above code, `processElement()` receives Transactions, and `processBroadcastElement()` receives Rule updates. When a new rule is created, it is distributed as depicted in Figure 6 and saved in all parallel instances of the operator using `processBroadcastState`. We use a Rule's ID as the key to store and reference individual rules. Instead of iterating over a hardcoded `List<Rules>`, we iterate over entries in the dynamically-updated broadcast state.
```